components/tikv_util/src/codec/stream_event.rs

```diff
@@ -1,13 +1,56 @@
 // Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+use crate::Either;
 use bytes::{Buf, Bytes};
 use std::io::prelude::*;
 use std::io::Cursor;
-use tikv_util::Either;
 
-pub struct Encoder;
+pub struct EventIterator {
+    buf: Cursor<Vec<u8>>,
+    index: usize,
+    len: usize,
+    pub val: Vec<u8>,
+}
+
+impl EventIterator {
+    pub fn new(buf: Vec<u8>) -> EventIterator {
+        let len = buf.len();
+        EventIterator {
+            buf: Cursor::new(buf),
+            index: 0,
+            len,
+            val: vec![],
+        }
+    }
+}
+
+impl Iterator for EventIterator {
+    type Item = Vec<u8>;
```

Comment:
why not return a `(Vec<u8>, Vec<u8>)`, but needing access to `self.val` to get the value

Reply:
I guess implement the iterator like RocksDB would be fine.

IMO by providing an internal buffer we can reduce some allocations. (Without https://github.com/rust-lang/rust/issues/44265, it seems that we cannot use `(&[u8], &[u8])` as the `Item` of iterator.) If we return `(Vec<u8>, Vec<u8>)`, this optimization would be harder to be done.

However the current method seems a little strange: we have implemented the std iterator but provide the value with rocksdb-style, maybe just make the iterator like this? 🤔

```rust
trait Iterator {
  fn next(&mut self) -> Result<()>;
  fn valid(&self) -> bool;
  fn key(&self) -> &[u8];
  fn value(&self) -> &[u8];
}
```

Reply:
okay, let me try

Reply:
the iterator trait has been updated. PTAL